airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+            else -> lock.read { flags[flag.key]?.enabled ?: flag.default }
+        }
+    }
+}
+
+/**
+ * Cloud feature-flag client. Feature-flags are derived from an external source (the LDClient).
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [client] the Launch-Darkly client for interfacing with Launch-Darkly.
+ */
+class CloudClient(private val client: LDClient) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
+            else -> client.boolVariation(flag.key, ctx.toLDUser(), flag.default)
+        }
+    }
+}
+
+/**
+ * Test feature-flag client. Intended only for usage in testing scenarios.
+ *
+ * All [Flag] instances will use the provided [values] map as their source of truth, including [EnvVar] flags.
+ *
+ * @param [values] is a map of [Flag.key] to enabled/disabled status.
+ */
+class TestClient(val values: Map<String, Boolean>) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> {
+                // convert to a EnvVar flag with a custom fetcher that uses the [values] of this Test class
+                // instead of fetching from the environment variables
+                EnvVar(envVar = flag.key, default = flag.default, team = flag.team) {
+                    values[flag.key]?.toString() ?: flag.default.toString()
+                }.enabled()
+            }
+
+            else -> values[flag.key] ?: flag.default
+        }
+    }
+}
+
+
+/**
+ * Data wrapper around OSS feature-flag configuration file.
+ *
+ * The file has the format of:
+ * flags:
+ *  - name: feature-one
+ *    enabled: true
+ *  - name: feature-two
+ *    enabled: false
+ */
+private data class PlatformFlags(val flags: List<PlatformFlag>)
+
+/**
+ * Data wrapper around an individual flag read from the configuration file.
+ */
+private data class PlatformFlag(val name: String, val enabled: Boolean)
+
+
+/** The yaml mapper is used for reading the feature-flag configuration file. */
+private val yamlMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+
+/**
+ * Reads a yaml configuration file, converting it into a map of flag name to flag configuration.
+ *
+ * @param [path] to yaml config file
+ * @return map of feature-flag name to feature-flag config
+ */
+//private fun readConfig(path: Path): Map<String, OSSFlag> = yamlMapper.readValue<List<OSSFlag>>(path.toFile())
+//    .associateBy { it.name }
+private fun readConfig(path: Path): Map<String, PlatformFlag> = yamlMapper.readValue<PlatformFlags>(path.toFile())
+    .let { it.flags }
+    .associateBy { it.name }
+
+/**
+ * Monitors a [Path] for changes, calling [block] when a change is detected.
+ *
+ * @receiver Path
+ * @param [block] function called anytime a change is detected on this [Path]
+ */
+private fun Path.onChange(block: () -> Unit) {
```

Comment:
If i understand right, this overrides behavior on the `Path` object. What is the scope of that override? Is it scoped just to this file?

Reply:
This doesn't override existing behavior on the `Path` object, it adds behavior to it.  Really it's just syntactical-sugar that Kotlin provides which roughly equates to the following java code:
```java
private static void onChange(final Path path, final Supplier<Void> block) {
  // ...
}
```

As this added behavior (aka extension function) is defined as `private`, it can only be called from within this file.  If it was defined as `internal` then it could be called from anywhere within the `airbyte-featureflag` module, and if defined as `public` (which is the default visibility if not defined) then it could be called from anywhere that this module was included as a dependency.

Reply:
Thanks! Makes sense.